// katana/template/app/config/development/stores.js
module.exports = {
  memory: {
    provider: 'memory'
  },

  /*redis: {
    provider: 'redis',

    host: '127.0.0.1',
    port: 6379,
    socket: null,
    password: null,
    database: null,

    options: {
      no_ready_check: false
    }
  },

  mongoose: {
    provider: 'mongoose',

    uri: null,

    host: '127.0.0.1',
    port: '27017',
    database: 'katana',

    options: {}
  }

  mongodb: {
    provider: 'mongodb',

    host: '127.0.0.1',
    port: 27017,
    database: 'katana',

    options: {
      server: {
        auto_reconnect: true
      },

      database: {}
    }
  },

  mysql: {
    provider: 'mysql',

    host: '127.0.0.1',
    port: 3306,
    username: 'root',
    password: '',
    database: 'katana'

    debug: false
  },

  postgre: {
    provider: 'postgre',

    host: '127.0.0.1',
    port: 5432,
    username: 'postgres',
    password: '',
    database: 'katana'
  }

  redis: {
    provider: 'redis',

    host: '127.0.0.1',
    port: 6379,
    socket: null,
    password: null,
    database: null,

    options: {
      no_ready_check: false
    }
  }*/
}
